staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,1010 @@
+// This is an alpha type and requires enabling the DynamicResourceAllocation
+// feature gate.
+type ResourceSlice struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Contains the information published by the driver.
+	//
+	// Changing the spec increments the generation number.
+	Spec ResourceSliceSpec `json:"spec" protobuf:"bytes,2,name=spec"`
+}
+
+// ResourceSliceSpec contains the information published by the driver in one ResourceSlice.
+type ResourceSliceSpec struct {
+	// Driver identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only ResourceSlice
+	// objects with a certain driver name.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	//
+	// +required
+	Driver string `json:"driver" protobuf:"bytes,1,name=driver"`
+
+	// Pool describes the pool that this ResourceSlice belongs to.
+	//
+	// +required
+	Pool ResourcePool `json:"pool" protobuf:"bytes,2,name=pool"`
+
+	// NodeName identifies the node which provides the devices.
+	// A field selector can be used to list only ResourceSlice
+	// objects belonging to a certain node.
+	//
+	// This field can be used to limit access from nodes to ResourceSlices with
+	// the same node name. It also indicates to autoscalers that adding
+	// new nodes of the same type as some old node might also make new
+	// devices available.
+	//
+	// Exactly one of NodeName and NodeSelector must be set.
+	//
+	// +optional
+	NodeName string `json:"nodeName,omitempty" protobuf:"bytes,3,opt,name=nodeName"`
+
+	// NodeSelector defines which nodes have access to the devices in the pool.
+	// If it is specified, but empty (has no terms), all nodes have access.
+	//
+	// Exactly one of NodeName and NodeSelector must be set.
+	//
+	// +optional
+	NodeSelector *v1.NodeSelector `json:"nodeSelector,omitempty" protobuf:"bytes,4,opt,name=nodeSelector"`
+
+	// Devices lists some or all of the devices in this pool.
+	//
+	// Must not have more than 128 entries.
+	//
+	// +optional
+	// +listType=atomic
+	Devices []Device `json:"devices" protobuf:"bytes,5,name=devices"`
+}
+
+// ResourcePool describes the pool that ResourceSlices belong to.
+type ResourcePool struct {
+	// Name is used to identify the pool. For node-local devices, this
+	// is often the node name, but this is not required.
+	//
+	// It must not be longer than 253 characters and must consist of one or more DNS sub-domains
+	// separated by slashes.
+	//
+	// +required
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// Generation gets incremented in all ResourceSlices of a pool whenever resource
+	// definitions change. A consumer must only use definitions from ResourceSlices
+	// with the highest generation number and ignore all others.
+	Generation int64 `json:"generation" protobuf:"bytes,2,name=generation"`
```

Comment:
> +required

Added.

> is this allowed to go backwards?

Not if there are existing slices. If all of them are deleted, then I think it's okay to start again at 0. There's a slight race where a scheduler might still use the "old" pool that already got deleted while a driver which right at that moment starts up, sees no slices, and starts publishing with `generation: 0`, but I think that's part of the unavoidable "scheduler might use out-dated information" problem.

The generation serves as signal for "pool is being updated", not for "pool information is the latest". We can indicate the former, but not the latter.

Added a paragraph about this.

>  is this allowed to increment by arbitrary jumps?

Yes. s/increment/increase/?

Reply:
I'm  probably not wording it well :cry: 

Resetting to zero is okay, so drivers don't need to remember any previous state. If they know about some existing slices, they should use a higher generation in their next update because it provides some early indication to the consumer that newer slices are coming while the older slices are being deleted.

Reply:
As a consumer, how do I implement the logic you want?  In particular, we said "use the latest generation".  If a driver can reset the generation to 0, how does that work?

IOW, my client has local state with 3 of 3 slices, all generation 123456.  Now I get an update that says one of those slices (or a new one) changed to generation 1.  You and I understand that the driver started over, but programmatically, how do I know that?

Reply:
Worded differently: a producer which publishes slices with generation 0 while leaving other, older slices with a higher generation in place is broken.

Reply:
A consumer cannot detect that kind of brokeness.

Reply:
Any issue with implementations just using a UTC time-based integer whenever an update is needed?

Reply:
Despite our reliance on clocks in other places, you can't really guarantee this is monotonic across node reboots, right?

I think I get what you are saying, @pohly.  Let me see if I can word it in a way that makes me happy, and then see if it doesn't corrupt your intention.

"""
Generation tracks the change in a pool over time.  Whenever a driver changes something about one or more of the resources in a pool, it must change the generation in all ResourceSlices which are part of that pool. Consumers of ResourceSlices should only consider resources from the pool with the highest generation number.  The generation may be reset by drivers, which should be fine for consumers, assuming that all ResourceSlices in a pool are updated to match or deleted.

Combined with ResourceSliceCount, this mechanism enables consumers to detect pools which are comprised of multiple ResourceSlices and are in an incomplete state.
"""

@pohly - Is it OK for a driver to set generation to 0?  Would it be a valid implementation for a driver to just oscillate between 1 and 2 on each subsequent set of updates?

Reply:
That wording sounds good to me. I'll ~~steal~~ use it.

>  Is it OK for a driver to set generation to 0? Would it be a valid implementation for a driver to just oscillate between 1 and 2 on each subsequent set of updates?

Both would be okay in the sense that eventually the pool is up-to-date again. But such a driver increases the risk that stale information is used.